trecs/components/users.py

```diff
@@ -24,18 +25,76 @@ def __init__(self, predicted_scores=None, verbose=False):
         )
 
+    def filter_by_index(self, item_indices):
+        """
+        Return a subset of the predicted scores, filtered by the indices
+        of valid items.
+
+        Parameters
+        -------------
+
+        item_indices: :obj:`numpy.ndarray` or `scipy.sparse.spmatrix`
+            A matrix with |U| rows that specifies the indices of items
+            requested for each user.
+
+        """
+        if item_indices.shape[0] != self.current_state.shape[0]:
+            error_msg = "Number of users does not match between score matrix and item index matrix"
+            raise ValueError(error_msg)
+        # generates row matrix like the following:
+        # [0, 0, 0, ..., 0]
+        # [1, 1, 1, ..., 1]
+        # [     ...       ]
+        # [n, n, n, ..., n]
+        num_users = item_indices.shape[0]
+        row = np.repeat(np.arange(num_users), item_indices.shape[1]).reshape((num_users, -1))
+        # for now, we have to keep the score matrix a dense array because scipy
+        # sparse has no equivalent of argsort
+        # TODO: look into potential solutions using things like Numba to maintain
+        # speed?
+        # https://stackoverflow.com/questions/31790819/scipy-sparse-csr-matrix-how-to-get-top-ten-values-and-indices
+        return mo.to_dense(self.current_state)[row, item_indices]
+
+    def append_new_scores(self, new_scores):
+        """
+        Appends a set of scores for new items to the current set of scores.
 
-class PredictedUserProfiles(Component):  # pylint: disable=too-many-ancestors
+        Parameters
+        -------------
+
+        new_scores: :obj:`numpy.ndarray` or `scipy.sparse.spmatrix`
+            Matrix of new scores with dimension :math:`|U|\\times|I_{new}|`,
+            where :math:`I_{new}` indicates the number of new items whose scores
+            are being to be appended.
+        """
+        self.current_state = mo.hstack([self.current_state, new_scores])
+
+
+class PredictedUsers(Component):  # pylint: disable=too-many-ancestors
     """
     User profiles as predicted by the model. This class is a container
     compatible with Numpy operations and it does not make assumptions on the
     size of the representation.
     """
 
     def __init__(self, user_profiles=None, size=None, verbose=False, seed=None):
-        self.name = "predicted_user_profiles"
+        self.name = "predicted_users"
         Component.__init__(self, current_state=user_profiles, size=size, verbose=verbose, seed=seed)
 
+    @property
+    def num_users(self):
+        """
+        Shortcut getter method for the number of users.
+        """
+        return self.current_state.shape[0]
+
+    @property
+    def num_attrs(self):
+        """
+        Shortcut getter method for the number of attributes in each user profile.
+        """
+        return self.current_state.shape[1]
+
 
 class ActualUserProfiles(Component):  # pylint: disable=too-many-ancestors
```

Comment:
Q: why did `PredictedUserProfiles` change to `PredictedUsers`, whereas `ActualUserProfiles` kept its name?

Reply:
mm, good point! I think it was because I added `PredictedItems` which made me want to change `PredictedUserProfiles` to `PredictedUsers`. Would you be in favor of changing `ActualUserProfiles` to `ActualUsers` as well?

Reply:
I am thinking the best solution could be to revert back to `PredictedUserProfiles` because I want to avoid confusion between `Users` and `ActualUsers` and still maintain the symmetry between `PredictedUserProfiles` and `ActualUserProfiles`. Thoughts?

Reply:
(not sure what the best solution is for items: I'd say I'm fine with `Items`-`PredictedItems` or `ItemAttributes`-`PredictedItemAttributes`. I do prefer the former because it's shorter and it can't be confused with anything else, but curious to hear your opinion.)

Reply:
Yes, happy to revert back to `PredictedUserProfiles`! That makes sense to me. Just pushed the change!